horovod/common/mpi_controller.cc

```diff
@@ -0,0 +1,215 @@
+// Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "mpi_controller.h"
+#include "logging.h"
+#include "operations.h"
+
+#if HAVE_GLOO
+#include "ops/gloo_operations.h"
+#endif
+
+namespace horovod {
+namespace common {
+
+// MPIController
+void MPIController::Initialize() {
+
+  // Check if multi-thread is supported.
+  int provided;
+  MPI_Query_thread(&provided);
+  mpi_threads_supported_ = (provided == MPI_THREAD_MULTIPLE);
+
+  // Get MPI rank to determine if we are rank zero.
+  MPI_Comm_rank(mpi_ctx_->mpi_comm, &rank_);
+  is_coordinator_ = rank_ == 0;
+
+  // Get MPI size to determine how many tensors to wait for before reducing.
+  MPI_Comm_size(mpi_ctx_->mpi_comm, &size_);
+
+  if (is_coordinator_) {
+    LOG(INFO) << "Started Horovod with " << size_ << " processes";
+  }
+
+  // Determine local rank by querying the local communicator.
+  MPI_Comm_rank(mpi_ctx_->local_comm, &local_rank_);
+  MPI_Comm_size(mpi_ctx_->local_comm, &local_size_);
+  local_comm_ranks_ = std::vector<int>((size_t)local_size_);
+  local_comm_ranks_[local_rank_] = rank_;
+  MPI_Allgather(MPI_IN_PLACE, 0, MPI_DATATYPE_NULL, local_comm_ranks_.data(), 1,
+                MPI_INT, mpi_ctx_->local_comm);
+
+  // Determine if cluster is homogeneous, i.e., if every node has the same
+  // local_size
+  local_sizes_ = std::vector<int>(size_);
+  MPI_Allgather(&local_size_, 1, MPI_INT, local_sizes_.data(), 1, MPI_INT,
+                mpi_ctx_->mpi_comm);
+
+  is_homogeneous_ = true;
+  for (int i = 0; i < size_; ++i) {
+    if (local_sizes_[i] != local_size_) {
+      is_homogeneous_ = false;
+      break;
+    }
+  }
+
+  // Get cross-node rank and size in case of hierarchical allreduce.
+  MPI_Comm_rank(mpi_ctx_->cross_comm, &cross_rank_);
+  MPI_Comm_size(mpi_ctx_->cross_comm, &cross_size_);
+}
+
+int MPIController::GetTypeSize(DataType dtype) {
+  return mpi_ctx_->GetMPITypeSize(dtype);
+}
+
+#if HAVE_GLOO
+void MPIController::InitializeGlooCtx(GlooContext& gloo_ctx) {
```

Comment:
This function actually can be moved to `gloo_context.cc` or `mpi_context.cc`, so that the independency between the controller and the context can be enforced.

Reply:
I like that!  Let's move it into `gloo_context.cc`.  In the next PR, we can wrap it in preprocessor checks to see if MPI is enabled.

Reply:
Moved.